src/engine/QueryExecutionTree.h

```diff
@@ -179,33 +191,44 @@ class QueryExecutionTree {
       }
       switch (validIndices[validIndices.size() - 1].second) {
         case ResultTable::ResultType::KB: {
-          string entity = _qec->getIndex().idToString(
+          std::optional<string> entity = _qec->getIndex().idToOptionalString(
               row[validIndices[validIndices.size() - 1].first]);
-          if (ad_utility::startsWith(entity, VALUE_PREFIX)) {
-            entity = ad_utility::convertIndexWordToValueLiteral(entity);
+          if (entity) {
+            string entitystr = entity.value();
+            if (ad_utility::startsWith(entitystr, VALUE_PREFIX)) {
+              entitystr = ad_utility::convertIndexWordToValueLiteral(entitystr);
+            }
+            os << "\"" << ad_utility::escapeForJson(entitystr) << "\"]";
+          } else {
+            os << "null]";
           }
-          os << ad_utility::escapeForJson(entity) << "\"]";
           break;
         }
         case ResultTable::ResultType::VERBATIM:
-          os << row[validIndices[validIndices.size() - 1].first] << "\"]";
+          os << "\"" << row[validIndices[validIndices.size() - 1].first]
+             << "\"]";
           break;
         case ResultTable::ResultType::TEXT:
-          os << ad_utility::escapeForJson(_qec->getIndex().getTextExcerpt(
+          os << "\""
+             << ad_utility::escapeForJson(_qec->getIndex().getTextExcerpt(
                     row[validIndices[validIndices.size() - 1].first]))
              << "\"]";
           break;
         case ResultTable::ResultType::FLOAT: {
           float f;
           std::memcpy(&f, &row[validIndices[validIndices.size() - 1].first],
                       sizeof(float));
-          os << f << "\"]";
+          os << "\"" << f << "\"]";
           break;
         }
         case ResultTable::ResultType::LOCAL_VOCAB: {
-          os << ad_utility::escapeForJson(res->idToString(
-                    row[validIndices[validIndices.size() - 1].first]))
-             << "\"]";
+          std::optional<string> entity = res->idToOptionalString(
+              row[validIndices[validIndices.size() - 1].first]);
+          if (entity) {
+            os << "\"" << ad_utility::escapeForJson(entity.value()) << "\"]";
+          } else {
+            os << "null]";
+          }
```

Comment:
I think it is not needed to introduce these duplications. You often have the workflow:
if optional value -> escape it for json (maybe convert to value)
else return "null";
This can perfectly be refactored into a function, with the "convert to value" (YES, NO, CHECK) being a parameter (runtime or template for efficiency).

Reply:
That function sounds oddly specific and it would also need to handle the quotes. I actually thought about changing `escapeForJson()` to take an `const std::optional<string>&` and rename it to something like `toJson()` but decided against it at the time to make it a less intrusive change. Then again that might actually be a good solution. It will still work for normal strings and it reflects that in JSON values can be null.

Reply:
ok, but the only thing we gain by your version, is that we know that "null]" does not need to be escaped. And I think there is not so much to gain there.
```
escapeForJson(entity.value_or("null")); as I see it.
```

Reply:
No my `toJson()` function (just pushed) also adds the quotes. So if the optional ist empty we get `"null"` otherwise `"\"escaped string\""`